### ISSUE TITLE
Use structured format for changelog entries

### ISSUE THREAD

**Issue by romac** [informalsystems/unclog#10]
Since virtually all our changelog entries follow the same format, I am wondering if we could add support to `unclog` for structured changelog entries, which would cut down a little bit on the "boilerplate".

Our entries typically follow this format:

**File:** `.changelog/unreleased/TYPE/ISSUE-SHORT_NAME.md`
**Content:**

```markdown
- Longer description of the change ([#1234])

[#1234]: https://github.com/informalsystems/unclog/issues/1234
```

Here are few suggestions:

a) Infer type and issue number from file path

**File:** `.changelog/unreleased/TYPE/ISSUE-SHORT_NAME.md`
**Content:**

```markdown
Longer description of the change
```

b) Structured format including type and issue number

**File:** `.changelog/unreleased/TYPE/ISSUE-SHORT_NAME.md`
**Content:**

```markdown
issue: 1234
type: bug-fix
description: Longer description of the change
```

c) Structured format including issue number only

**File:** `.changelog/unreleased/TYPE/ISSUE-SHORT_NAME.md`
**Content:**

```yaml
issue: 1234
description: Longer description of the change
```

For b) and c) we need to pick a suitable format, eg. YAML/TOML/JSON/custom.

Additionally, we could add support for templates, eg. by creating a `.changelog/template.md` file (with made-up format below):

```markdown
- ${description} ([#${issue.number}](${issue.link})) by @${author.github}
```

What do you think?

**Comment by thanethomson** [comment]
It's definitely something I'm considering. I'm also trying to be mindful of all the potential complexities that could arise from adding more structure.

Up to now, I've been trying to avoid going down the route of offering any more structure to the format than what I've already implemented. I've even been trying really really hard to avoid having to have a configuration file. Largely because structured formats make assumptions about how people are going to use the tool, which, especially early in the development and use of a tool, tend to become restrictive and difficult to backtrack on. This inevitably ends up sending you down to configuration hell, where you have to continuously tweak your structure to accommodate a wider variety of tastes and preferences and edge cases (especially the edge cases). Sometimes people start inventing DSLs and even embedding crude programming languages into their formats to be able to cater for this variety.

Already your suggestion in (c), for example, along with the templating system, could actually be accomplished with a number of static content generators. (I even wrote one [here](https://github.com/thanethomson/statik) that could probably accomplish this - it needs some love, but could do the job.)

For example, how would you cater for the hierarchical structure differences between the [`tendermint-rs` changelog](https://github.com/informalsystems/tendermint-rs/blob/master/CHANGELOG.md) and the [`ibc-rs` changelog](https://github.com/informalsystems/ibc-rs/blob/master/CHANGELOG.md) formats in how issues are grouped by crate? Would you need multiple levels of templates to accomplish this? (i.e. one for the overall changelog, and another for groupings of issues like in `ibc-rs`, and yet another for individual issues?) And how many other structured formats would users potentially want? And at what point does `unclog` cease to be an opinionated, small, easy-to-tweak changelog management tool that *just works<sup>TM</sup>* and evolve into a more complex, generic, templated content transformation system that requires reading a lot more documentation and a steeper learning curve before you can use it productively?

It's hard to find the right balance there. Again, we should definitely consider different structured approaches, but as for which one would facilitate the most ease of use for developers, I'm not sure yet. More usage is probably needed :grin: 

Also, as a side note, the format I'd recommend using for issues right now is one that uses embedded links rather than the referenced structure in your example, like so:

```markdown
- Longer description of the change ([#1234](https://github.com/informalsystems/unclog/issues/1234))
```

I was thinking of at least building in a slightly more intelligent Markdown parser that would:

1. Automatically inject bullet points into the entries.
2. Automatically extract issue numbers from filenames.
3. Automatically inject the hyperlinks for issue numbers.

So you'd be able to simply add an entry `.changelog/unreleased/TYPE/1234-issue-slug.md`, with content:

```markdown
Longer description of the change
```

And it'd automatically be transformed into the previous example. Even that would probably eventually require a configuration file to facilitate. But I'd want to reduce the amount of information duplication down to an absolute minimum (e.g. if there's info in a filename, we shouldn't have to include it in the Markdown content too).

**Comment by thanethomson** [comment]
I think that the newly released v0.4.0 effectively fulfills this requirement. Please see the [README](https://github.com/informalsystems/unclog/blob/master/README.md) for the new usage info and let me know what's still missing for your use case?

**Comment by romac** [comment]
Wow, that's amazing and does indeed fulfill my wishes. Thanks a lot @thanethomson :)
